GitHub Release publishing tweaks

MattSturgeon opened the pull request:
Motivated by #7, introduce some basic control over how GitHub Releases are published.

- Allow configuring whether missing/existing releases are created/updated
- Allow publishing to a tag that is different to `version`

I'm fairly unfamiliar with your codebase, so feel free to propose changes, push commits here, or discard this PR in favour of your own implementation.

MattSturgeon committed:
[FEAT]: Decouple GitHub release tag from version

Introduce `githubTag` property, which defaults to `version` but can be configured separately.

MattSturgeon committed:
[FEAT]: Configure existing/missing GH releases

Add `createGithubRelease` and `updateGithubRelease` properties that can
be disabled to prevent publishing to a GitHub release in the respective
scenarios.

hypherionmc commented on src/main/java/com/hypherionmc/modpublisher/tasks/GithubUploadTask.java:
```diff
@@ -86,19 +86,22 @@ public void upload() throws Exception {
             if (extension.getDisplayName().isPresent() && !extension.getDisplayName().get().isEmpty()) {
                 releaseBuilder.name(extension.getDisplayName().get());
             } else {
+                // TODO make version optional if tag is set
```
Since tag defaults to version, would it make sense to just pass tag here, instead of hard pulling it from version?

MattSturgeon replied:
IDK. Usually a name would be defined, but if not would a user expect the tag or the version to be used (assuming they're different)?

hypherionmc replied:
The version fallback was just added here, since afaik, GitHub does require a name for releases? I could be wrong about this.

Maybe it can be a case of using the display name, or tag if it differs from the version. Idk, or otherwise this should be another config setting.

MattSturgeon replied:
> Maybe it can be a case of using the display name, or tag if it differs from the version

Well since `tag` defaults to `version`, it will always be defined.

`version` could be in any of three states; undefined, matching, or different.

The main reason for `tag` to exist in the first place is to allow `version` to be different, but if a user only cares about GitHub and isn't publishing to Modrinth, `version` could be optional...

My `TODO` comment had something like `version().getOrDefault(tag)` in mind. Does that work for you?

hypherionmc replied:
> version().getOrDefault(tag)

This sounds good to me. Plain and simple

MattSturgeon replied:
Done

hypherionmc commented on src/main/java/com/hypherionmc/modpublisher/tasks/GithubUploadTask.java:
```diff
@@ -96,6 +96,11 @@ public void upload() throws Exception {
         // Existing release was not found, so we create a new one
         if (ghRelease == null) {
+             if (!extension.getCreateGithubRelease().get()) {
+                 project.getLogger().warn("Create GitHub Release is disabled and Github Release with tag " + tag + " does not exist");
```
Instead of a warn here, would it make more sense to hard fail, since this would technically be an error?

MattSturgeon replied:
> since this would technically be an error

Had the same thought process but came to the opposite conclusion :joy: 

IMO this would be configurable too, since some users will want to fail while others would want to "attempt" to publish, unless exists/missing

hypherionmc replied:
That's fair. Might as well just leave it at a warning lol

hypherionmc commented on src/main/java/com/hypherionmc/modpublisher/tasks/GithubUploadTask.java:
```diff
@@ -109,6 +114,9 @@ public void upload() throws Exception {
             releaseBuilder.commitish(ghRepository.getDefaultBranch());
             ghRelease = releaseBuilder.create();
+        } else if (!extension.getUpdateGithubRelease().get()) {
+            project.getLogger().warn("Update GitHub Release is disabled and Github Release with tag " + tag + " already exists");
```
Same here. Should we hard fail?

MattSturgeon commented:
I think we should also add a `createGithubTag` option, checks if the [tag exists](https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28#get-a-tag) before creating a release.

This would be similar to the `gh` CLI's [`--verify-tag` option](https://cli.github.com/manual/gh_release_create).

hypherionmc commented:
> I think we should also add a `createGithubTag` option, checks if the [tag exists](https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28#get-a-tag) before creating a release.
> 
> This would be similar to the `gh` CLI's [`--verify-tag` option](https://cli.github.com/manual/gh_release_create).

Sounds good. Could be useful to add

MattSturgeon committed:
[FEAT]: Configure existing/missing GH tags

MattSturgeon commented:
All points addressed.

Let me know if there's any other changes, or if you'd like any of these options included in the testprojects.

It'd also be nice to expose configuring whether the release is a draft, and the "commitish" used when creating a new tag. Perhaps in a future PR...

MattSturgeon commented:
How should multi-project builds play into this?

As is, if a user configured `updateGithubRelease = false`, for example, only one of their projects would publish. It'd be undefined which one, too...

We could document this somewhere. Unless there's a better approach where some checks could be shared across projects or something?

If this is opening up a can of worms, maybe we scale this PR down to just `githubReleaseTag` and postpone the boolean properties?

EDIT: I'm wondering if exposing some function properties would be a better approach.

E.g. `shouldCreateRelease(ctx)` and `shouldUpdateRelease(ctx)`. Having a context object would allow the user to implement custom logic.

Similar provider functions could be used for the updated values, such as release body, display name, etc... 🤔

hypherionmc commented:
Maybe it could make more sense to have a dedicated GitHub section, similar to what we have now for overriding the values of the additional files on curse.

The behavior would then be to use the default configuration as before, but anything under that section would override the default values.

Multiproject usage shouldn't be a major issue, as the plugin is usually configured per project, rather than once in root.

MattSturgeon committed:
[REFACTOR]: Nest github options in their own DSL

MattSturgeon committed:
[REFACTOR]: Move `githubRepo` to `github.repo`

Not deprecating the old name yet, because it is consistent with `curseID` and `modrinthID`.

MattSturgeon commented:
> Maybe it could make more sense to have a dedicated GitHub section

Agreed, I've attempted that in my latest push. Not sure if it's correct or not though, wasn't able to test.

MattSturgeon committed:
Merge branch '2.0'

# Conflicts:
#	src/main/java/com/hypherionmc/modpublisher/plugin/ModPublisherGradleExtension.java

hypherionmc commented:
I'll take a look over everything tonight. Day is insanely crazy. Thank you :)
